src/framework/ui/internal/platform/windows/windowsplatformtheme.cpp

```diff
@@ -107,26 +80,83 @@ bool WindowsPlatformTheme::isFollowSystemThemeAvailable() const
     return m_buildNumber >= 17763; // Dark theme was introduced in Windows 1809
 }
 
-bool WindowsPlatformTheme::isDarkMode() const
+std::string WindowsPlatformTheme::bestSuitedThemeCode() const
+{
+    if (isSystemHighContrast()) {
+        return HIGH_CONTRAST_THEME_CODE;
+    }
+
+    if (isSystemDarkMode()) {
+        return DARK_THEME_CODE;
+    }
+
+    // Don't use high contrast theme, because it is too dark
+    // Light high contrast theme would be nice
+    return LIGHT_THEME_CODE;
+}
+
+Channel<std::string> WindowsPlatformTheme::bestSuitedThemeCodeChanged() const
+{
+    return m_channel;
+}
+
+bool WindowsPlatformTheme::isSystemDarkMode() const
 {
     DWORD data {};
     DWORD datasize = sizeof(data);
-    if (RegGetValue(HKEY_CURRENT_USER, windowsThemeKey.c_str(), windowsThemeSubkey.c_str(),
+    if (RegGetValue(HKEY_CURRENT_USER, windowsThemesPersonalizeKey.c_str(), windowsThemesPersonalizeSubkey.c_str(),
                     RRF_RT_REG_DWORD, nullptr, &data, &datasize) == ERROR_SUCCESS) {
         return data == 0;
     }
     return false;
 }
 
-Channel<bool> WindowsPlatformTheme::darkModeSwitched() const
+bool WindowsPlatformTheme::isSystemHighContrast() const
 {
-    return m_darkModeSwitched;
+    HIGHCONTRAST info;
+    info.cbSize = sizeof(HIGHCONTRAST);
+    if (SystemParametersInfoW(SPI_GETHIGHCONTRAST, 0, &info, 0)) {
+        return info.dwFlags & HCF_HIGHCONTRASTON;
+    }
+    return false;
+}
+
+void WindowsPlatformTheme::th_listen()
+{
+    static const DWORD dwFilter = REG_NOTIFY_CHANGE_NAME
+                                  | REG_NOTIFY_CHANGE_ATTRIBUTES
+                                  | REG_NOTIFY_CHANGE_LAST_SET
+                                  | REG_NOTIFY_CHANGE_SECURITY;
+
+    while (m_isListening) {
+        hStopListeningEvent = CreateEvent(NULL, FALSE, TRUE, TEXT("StopListening"));
+        hThemeChangeEvent = CreateEvent(NULL, FALSE, TRUE, TEXT("ThemeSettingChange"));
+        if (RegNotifyChangeKeyValue(hKey, TRUE, dwFilter, hThemeChangeEvent, TRUE) == ERROR_SUCCESS) {
+            std::string oldBestSuited = bestSuitedThemeCode();
+            HANDLE handles[2] = { hStopListeningEvent, hThemeChangeEvent };
+            DWORD dwRet = WaitForMultipleObjects(2, handles, FALSE, 4000);
+            if (dwRet != WAIT_TIMEOUT && dwRet != WAIT_FAILED) {
+                if (!m_isListening) {
+                    // then it must have been a stop event
```

Comment:
Better don't use empty if-else blocks. It will be simpler if you rewrite your code like this:

if (m_isListening) {
Sleep(100); // There might be some delay
std::string newBestSuited = bestSuitedThemeCode();
....
}

Reply:
Thanks, fixed!